homeassistant/components/volkswagencarnet/__init__.py

```diff
@@ -0,0 +1,237 @@
+
+            data.instruments.add(instrument)
+            discovery.load_platform(hass, COMPONENTS[instrument.component], DOMAIN, (vehicle.vin,instrument.component,instrument.attr), config)
+
+    def update(now):
+        """Update status from Volkswagen Carnet"""
+        try:
+            # check if we can login again
+            if not connection.logged_in:
+                connection._login()
+                if not connection.logged_in:
+                    _LOGGER.warning('Could not login to volkswagen carnet, please check your credentials')
+                    return False
+            else:
+                if not connection.update(request_data = False):
+                    _LOGGER.warning("Could not query update from volkswagen carnet")
+                    return False
+                else:
+                    _LOGGER.debug("Updating data from volkswagen carnet")
+
+                    for vehicle in connection.vehicles:
+                        if vehicle.vin not in data.vehicles:
+                            _LOGGER.info("Adding data for VIN: %s from carnet" % vehicle.vin.lower())
+                            discover_vehicle(vehicle)
+                        
+                        for entity in data.entities[vehicle.vin]:
+                            entity.schedule_update_ha_state()
+
+                        dispatcher_send(hass, SIGNAL_STATE_UPDATED, vehicle)
+            return True
+
+        finally:
+            track_point_in_utc_time(hass, update, utcnow() + interval)
+
+    _LOGGER.info("Starting volkswagencarnet component")
+    return update(utcnow())
+
+
+class VolkswagenData:
+    """Hold component state."""
+
+    def __init__(self, config):
+        """Initialize the component state."""
+        self.vehicles = set()
+        self.instruments = set()
+        self.entities = {}
+        self.config = config[DOMAIN]
+        self.names = self.config.get(CONF_NAME)
+
+    def instrument(self, vin, component, attr):
+        """Return corresponding instrument."""
+        return next((instrument
+                     for instrument in self.instruments
+                     if instrument.vehicle.vin == vin and
+                     instrument.component == component and
+                     instrument.attr == attr), None)
+
+    def vehicle_name(self, vehicle):
+        """Provide a friendly name for a vehicle."""
+        if (vehicle.vin and vehicle.vin.lower() in self.names):
+            return self.names[vehicle.vin.lower()]
+        elif vehicle.vin:
+            return vehicle.vin
+        else:
+            return ''
+
+
+class VolkswagenEntity(Entity):
+    """Base class for all Volkswagen entities."""
+
+    def __init__(self, data, vin, component, attribute):#, attribute):
+        """Initialize the entity."""
+        self.data = data
+        self.vin = vin
+        self.component = component
+        self.attribute = attribute
+
+        self.data.entities[self.vin].append(self)
```

Comment:
Don't store the entity in a shared container. Use the dispatch helper as shown above.